ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTable.scala

```diff
@@ -6,486 +6,275 @@ package com.digitalasset.platform.store.dao.events
-          |  {is_root},
-          |  {command_id},
-          |  {application_id},
-          |  {submitter},
-          |  {exercise_consuming},
-          |  {exercise_choice},
-          |  {exercise_argument},
-          |  {exercise_result},
-          |  {exercise_actors},
-          |  {exercise_child_event_ids}
-          |)
-          |""".stripMargin
+      // Get all the visible events in this transactions to filter children in each event
+      val visible = events.iterator.map(_.event.eventId).toSet
 
-  private def exercise(
-      applicationId: Option[ApplicationId],
-      workflowId: Option[WorkflowId],
-      commandId: Option[CommandId],
-      transactionId: TransactionId,
-      nodeId: NodeId,
-      submitter: Option[Party],
-      roots: Set[NodeId],
-      ledgerEffectiveTime: Date,
-      offset: Offset,
-      exercise: Exercise,
-  ): Vector[NamedParameter] =
-    Vector[NamedParameter](
-      "event_id" -> fromTransactionId(transactionId, nodeId).toString,
-      "event_offset" -> offset,
-      "contract_id" -> exercise.targetCoid.coid.toString,
-      "transaction_id" -> transactionId.toString,
-      "workflow_id" -> workflowId.map(_.toString),
-      "ledger_effective_time" -> ledgerEffectiveTime,
-      "template_package_id" -> exercise.templateId.packageId.toString,
-      "template_name" -> exercise.templateId.qualifiedName.toString,
-      "node_index" -> nodeId.index,
-      "is_root" -> roots(nodeId),
-      "command_id" -> commandId.map(_.toString),
-      "application_id" -> applicationId.map(_.toString),
-      "submitter" -> submitter.map(_.toString),
-      "exercise_consuming" -> exercise.consuming,
-      "exercise_choice" -> exercise.choiceId.toString,
-      "exercise_argument" -> serializeExerciseArgOrThrow(exercise),
-      "exercise_result" -> serializeNullableExerciseResultOrThrow(exercise),
-      "exercise_actors" -> exercise.actingParties.map(_.toString).toArray,
-      "exercise_child_event_ids" -> exercise.children
-        .map(fromTransactionId(transactionId, _): String)
-        .toArray,
-    )
+      // All events in this transaction by their identifier, with their children
+      // filtered according to those visible for this request
+      val eventsById =
+        events.iterator.map(r => r.event.eventId -> r.event.filterChildEventIds(visible)).toMap
 
-  private val updateArchived =
-    """update participant_events set create_consumed_at={consumed_at} where contract_id={contract_id}"""
+      // Roots are all the visible events in this transaction
+      // that don't appear in any event's children
+      val rootEventIds =
+        visible.diff(eventsById.valuesIterator.flatMap(_.childEventIds).toSet).toSeq
 
-  private def archive(
-      contractId: ContractId,
-      consumedAt: Offset,
-  ): Vector[NamedParameter] =
-    Vector[NamedParameter](
-      "consumed_at" -> consumedAt,
-      "contract_id" -> contractId.coid.toString,
-    )
+      (eventsById, rootEventIds)
```

Comment:
Yes, thanks.

Reply:
Since I'm already ordering by `node_index` on the database I'll change `rootEventIds` to preserve the order and use that.

Reply:
Since I'm retrieving items sorted by node index from the database, I'm writing a solution in which I'm simply preserving the original order (for `rootEventIds`).

Reply:
Addressed by https://github.com/digital-asset/daml/pull/5089/commits/267c5948c2d99274b1a7c3baa15e2f98b5d6b884